rdtools/degradation.py

```diff
@@ -80,3 +80,293 @@ def degradation_with_ols(normalized_energy):
+    try:
+        energy_median = normalized_energy.median()
+        dataframe = dataframe.resample('MS').mean()
+        
+        if interpolate_flag:
+            dataframe = dataframe.interpolate()    
+        else:  #append the median value to missing months
+            dataframe = dataframe.fillna(value=energy_median)
+        
+        dataframe['Month'] = np.arange(0, len(dataframe))
+
+    except:
+        # if you don't pass a datetime index, it assumes that there is data for each month
+        dataframe['Month'] = np.arange(0, len(dataframe))
+        
+    if pd.infer_freq(dataframe.index) == 'MS' and len(normalized_energy) > 12:
+        dataframe = dataframe.rolling(window = 12, center = True).mean()
+        
+    dataframe = dataframe.dropna()
+    y2=dataframe['normalized_energy']
+    x2 = dataframe['Month']
+    
+
+
+    if len(dataframe) ==0:
+        degradation_values = {}
+
+        print '\nNot enough data for seasonal decomposition:'
+        
+        
+    else:        
+        # OLS regression with constant
+        x2=sm.add_constant(x2)
+        model2 = sm.OLS(y2,x2).fit()
+    
+        b_cd = model2.params['const']
+        m_cd = model2.params['Month']
+        SE_b_cd = model2.bse['const']
+        SE_m_cd = model2.bse['Month']
+    
+        Rd_cd = (m_cd * 12) / b_cd * 100
+        SE_Rd_cd = np.sqrt(np.power(SE_m_cd * 12/b_cd, 2) + np.power((-12*m_cd/b_cd**2) * SE_b_cd, 2))*100
+
+        print '\nDegradation and Standard Error of Classical decomposition:'
+        print 'Rd = {:.2f} +/- {:.2f}'.format(Rd_cd, SE_Rd_cd)  
+
+        degradation_values = {
+        'Rd_pct': Rd_cd,
+        'slope': m_cd,
+        'intercept': b_cd,
+        #'rmse': rmse,
+        #'slope_stderr': stderr_m,
+        #'intercept_stderr': stderr_b,
+        'Rd_stderr_pct': SE_Rd_cd,
+        'Dataframe':dataframe
+        }
+    return(degradation_values)
+     
+
+
+def degradation_year_on_year(normalized_energy):
+    # (YOY_median,YOY_unc,YoY_filtered1) = degradation_year_on_year(normalized_energy)
+    #
+    # year-on-year  decomponsition method 
+    # from Dirk Jordan (NREL) 12/10/16
+    # input parameters:  
+    #       series['normalized_energy'] : corrected performance ratio series
+    #       index:  timeseries index in monthly format
+    #       
+    #output parameters:
+    #   (YOY_median,YOY_unc) : tuple of median degradation rate and standard error
+    # TODO:  check timebase for other input time options (day, week)    
+
+    ################################################################################################################################
+    # YOY approach
+    
+    # check the index frequency
+    freq = pd.infer_freq(normalized_energy.index)
+    if freq == 'MS': #monthly
+        YearSampleSize = 12
+    elif freq == 'W':
+        YearSampleSize = 52
+    elif freq == 'D':
+        YearSampleSize = 365
+    else:  #sample to daily by default
```

Comment:
For debugging purposes, it might be best to explicitly name frequencies under a day and throw an exception if the frequency is not supported, something like:

```
elif freq in ['H', '30T', '15T', 'T']:
    # resample to a daily
else:
    raise Exception('Frequency {} not supported'.format(freq))
```